### ISSUE TITLE
Background index cannot be enabled when using "compile_flags.txt"

### ISSUE THREAD

**Issue by lh123** [clangd/clangd#82]
Background index cannot be enabled when using "compile_flags.txt"

**Comment by kadircet** [comment]
Yes, background index currently requires underlying compilation database to list all the files in the project. 

Unfortunately only JSONCompilationDatabase implements that functionality as of today, but I think it makes sense to perform discovery in the case of FixedCompilationDatabase, since we have the compile commands for any files we've found.

Only question is how to decide on what looks like a source code, because clang might crash if we run it on arbitrary files, and that is catastrophic for background indexer.

**Comment by lh123** [comment]
> Only question is how to decide on what looks like a source code, because clang might crash if we run it on arbitrary files, and that is catastrophic for background indexer.

The "vscode-cpptools" and "ccls" seems to distinguish whether it is a source file based on the suffix name.

**Comment by sam-mccall** [comment]
Yeah. I think we've been averse to that option because it wouldn't work well for the clang/clangd/llvm code base.

Also there's some conflicting design goals here:
 - clangd doesn't have an explicit workspace, the workspace is inferred from the containing CDB
 - you can drop e.g. `-Wall` in ~/compile_flags.txt and it'll be the default of any files in your home directory
 - but indexing all code recursively inside your home directory is probably a bad idea

We'd likely need some fancy heuristics...

**Comment by kadircet** [comment]
Closing this as we are unlikely to prioritize it as there's no simple solution that'll make it work in general without providing suboptimal UX (due to reasons mentioned above) and it's a rare use case.

For future travelers, feel free to re-open if you got some ideas/heuristics around how this might be tackled in a way that won't introduce too much complexity into the common parts of the codebase and would provide a sensible UI. I'll also put this under a known limitations sections to make it more visible.
